docs/examples/langchain/langchain_chat_pandas_df.py

```diff
@@ -0,0 +1,185 @@
+
+    show_chain_of_thought = param.Boolean(default=False)
+
+    def _get_agent_message(self, message: str) -> pn.chat.ChatMessage:
+        return pn.chat.ChatMessage(message, user=self.user, avatar=self.avatar)
+
+
+class AppState(param.Parameterized):
+    data = param.DataFrame()
+
+    llm = param.Parameter(constant=True)
+    pandas_df_agent = param.Parameter(constant=True)
+
+    config: AgentConfig = param.ClassSelector(class_=AgentConfig)
+    environ: Environment = param.ClassSelector(class_=Environment, constant=True)
+
+    def __init__(
+        self, config: AgentConfig | None = None, environ: Environment | None = None
+    ):
+        if not config:
+            config = AgentConfig()
+        if not environ:
+            environ = Environment()
+        super().__init__(config=config, environ=environ)
+
+    @param.depends("environ.OPENAI_API_KEY", on_init=True, watch=True)
+    def _reset_llm(self):
+        with param.edit_constant(self):
+            if self.environ.OPENAI_API_KEY:
+                self.llm = ChatOpenAI(
+                    temperature=0,
+                    model="gpt-3.5-turbo-0613",
+                    api_key=self.environ.OPENAI_API_KEY,
+                    streaming=True,
+                )
+            else:
+                self.llm = None
+
+    @param.depends("llm", "data", on_init=True, watch=True)
+    def _reset_pandas_df_agent(self):
+        with param.edit_constant(self):
+            if not self.error_message:
+                self.pandas_df_agent = create_pandas_dataframe_agent(
+                    self.llm,
+                    self.data,
+                    verbose=True,
+                    agent_type=AgentType.OPENAI_FUNCTIONS,
+                    handle_parsing_errors=True,
+                )
+            else:
+                self.pandas_df_agent = None
+
+    @property
+    def error_message(self):
+        if not self.llm and self.data is None:
+            return dedent(
+                """\
+                Please provide your `OPENAI_API_KEY`, **upload a `.csv` file**
+                and click the **send** button."""
+            )
+        if not self.llm:
+            return "Please provide your `OPENAI_API_KEY`."
+        if self.data is None:
+            return "Please **upload a `.csv` file** and click the **send** button."
+        return ""
+
+    @property
+    def welcome_message(self):
+        text = dedent(
+            f"""
+            I'm your <a href="\
+            https://python.langchain.com/docs/integrations/toolkits/pandas" \
+            target="_blank">LangChain Pandas DataFrame Agent</a>.
+
+            I execute LLM generated Python code under the hood - this can be bad if
+            the `llm` generated Python code is harmful. Use cautiously!
+
+            {self.error_message}"""
+        ).strip()
+        if self.data is None:
+            text += f"""
+
+Example: <a href="{PENGUINS_URL}" download>penguins.csv<a>
```

Comment:
To make this easier on the user, can this be as FileDownload widget? When I click this link, it just opens a new tab rather than download.

`instance.send(pn.Column("You can download an example file below", pn.widgets.FileInput(...)))`

Reply:
Done. Its not without costs though.

1) The file has to be available locally. I.e. making it impossible to copy paste the code only 2) The FileDownload button is very huge, so in order for it not to take all the focus I have to add a complicated `stylesheet`.

Reply:
Regarding 1, you could theoretically add `requests.get()` if not exists.

Maybe instead of stylesheet, maybe change `design` to another template style, or call  `send` so it's a separate message.